Only change rootmultistore hash when substore hashes change

ethanfrey opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

Closes #1351 

The fix was proposed one year ago, and was recently moved to "proposal accepted". This enables no-empty-block to be enabled with cosmos-sdk code without any other breaking changes.

- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/develop/CONTRIBUTING.md#pr-targeting))

- [x] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [x] Wrote tests
- [x] Updated relevant documentation (`docs/`)
- [x] Added a relevant changelog entry: `clog add [section] [stanza] [message]`
- [x] rereviewed `Files changed` in the github PR explorer

Note: I don't know of any relevant documentation as this is a bugfix. Please point me to any points I need to update if any.
______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

ethanfrey commented:
Please note that this issue was brought up as a problem by some groups during the Hackatom Berlin.

Also, this not being resolved has prompted other quick fixes, which would have much more breaking impacts, such as https://github.com/tendermint/iavl/pull/143

ethanfrey commented:
> Concept ACK. I'm pretty sure this is state machine breaking, this I've added the label. Non-contentious update regardless. Thanks @ethanfrey +1

Yes, this will break all app hashes and require a chain restart for an upgrade to new version. I am fine if this merge waits until there are other state-machine breaking changes lined up, to all go in one release.

alexanderbez commented:
@ethanfrey we have a significant amount of breaking changes already in `master` and ready for the next major upgrade. I see no reason we can't get this in now :)

alexanderbez commented:
Suggestions on a location @cwgoes?

cwgoes commented:
> Suggestions on a location @cwgoes?

In the default configuration file and in comments above `EndBlock` in the module abstraction perhaps.

mdyring commented:
Thanks for the work on this @ethanfrey - much appreciated!

ethanfrey commented:
> Concept ACK, code utACK

What does `utACK` mean?

I will add comments about `no_empty_blocks` to the two suggested locations later on today.

cwgoes commented:
> What does utACK mean?

Acknowledged (read the code, looks correct) but untested.

jackzampolin commented:
Is this something we want to get in for gaia 1.0?

alexanderbez commented:
@jackzampolin yes. I'll be merging this once @ethanfrey updates some docs.

ethanfrey commented:
I rebased on master and added some comments to `BeginBlocker`/`EndBlocker` interfaces, which will be used be developers.

`baseapp.BaseApp.{Begin,End}Block` seemed like large code blocks that only core devs look at, so I didn't add docs there (but could if you want).

I could not find the default configuration besides `client/config.go` with no mention of `no_empty_blocks` in any toml file in the database. I assume this comes from the tendermint default toml file and the comment would have to go there:

https://github.com/tendermint/tendermint/blob/4253e67c07c69be6d7f7263ab03944ce30a9fc90/config/toml.go#L294-L297

(I also see that the name is now `create_empty_blocks` and will update the docs accordingly)

ethanfrey commented:
I just found this PR: https://github.com/cosmos/cosmos-sdk/pull/2284/files

I guess I should revert that as well. Does that belong in this PR or another one?

alexanderbez commented:
> I just found this PR: https://github.com/cosmos/cosmos-sdk/pull/2284/files
>
> I guess I should revert that as well. Does that belong in this PR or another one?

Yes, let's create another PR for that too.

mostcute commented:
Hey friends,I want to support create-empty-block=false but i don't want to update my sdk version. how many changes i have to merge to my code? my sdk version is 0.34.4 
I have merge the diff in #4613 &&  #2284.
and set the config.toml as
```
timeout_commit = "1s"

# Make progress as soon as we have all the precommits (as if TimeoutCommit = 0)
skip_timeout_commit = false

# EmptyBlocks mode and possible interval between empty blocks
create_empty_blocks = false
create_empty_blocks_interval = "10s"
```
But my program still product block per second
